core/src/main/scala/org/apache/spark/sql/execution/columnar/impl/JDBCSourceAsColumnarStore.scala

```diff
@@ -370,42 +486,41 @@ class JDBCSourceAsColumnarStore(override val connProperties: ConnectionPropertie
      }
   }
 
-  protected def getPartitionID(tableName: String,
+  // use the same saved connection for all operation
+  private def getPartitionID(conn: Option[Connection], tableName: String,
       partitionId: Int = -1): Int = {
-    val connection = getConnection(tableName, onExecutor = true)
-    try {
-      connectionType match {
-        case ConnectionType.Embedded =>
-          val resolvedName = ExternalStoreUtils.lookupName(tableName,
-            connection.getSchema)
-          val region = Misc.getRegionForTable(resolvedName, true)
-              .asInstanceOf[LocalRegion]
-          region match {
-            case pr: PartitionedRegion =>
-              if (partitionId == -1) {
-                val primaryBucketIds = pr.getDataStore.
-                    getAllLocalPrimaryBucketIdArray
-                // TODO: do load-balancing among partitions instead
-                // of random selection
-                val numPrimaries = primaryBucketIds.size()
-                // if no local primary bucket, then select some other
-                if (numPrimaries > 0) {
-                  primaryBucketIds.getQuick(rand.nextInt(numPrimaries))
-                } else {
-                  rand.nextInt(pr.getTotalNumberOfBuckets)
-                }
+    val connection = conn match {
+      case Some(conn) => conn
+      case None => getConnection(tableName, onExecutor = true)
```

Comment:
Here we are passing the already used connection. If not passed i.e None, then gets a new pooled connection.